### PR TITLE
Refactor llama / mixtral / grok for shared features

### PR DESCRIPTION
**Pull request by rsuderman**
Many of these features can toggle between depending on architecture. Replumbing the configurations separately allows better reuse and understanding of how models vary between eachother.

grok uses a softcap, plumbing a value enables `sc * tanh( v / sc)` grok has some hardcoded values that have better representations, e.g. `sqrt(6144)` and `sqrt(3)`.

output normalization is optional but used by mixtral. Presence of the tensor is sufficient for performing the normalization.

### CODE REVIEW

File: sharktank/sharktank/models/grok/grok.py [hunk 1]
```diff
@@ -93,16 +95,15 @@ def __init__(self, theta: Theta, config: LlamaModelConfig):
...
-            self.attn_blocks.append(
-                PreGatherMoeBlock(
+            self.moe_blocks.append(
+                MoeBlock(
                     theta("blk", n),
                     expert_count=hp.expert_count,
                     expert_used_count=hp.expert_used_count,
                     rms_epsilon=hp.attention_layer_norm_rms_epsilon,
-                    use_grok=True,
+                    activation=F.gelu,
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/grok/grok.py:106]: Should this be `moe_activation=F.gelu` as defined in the `MoeBlock`?

File: sharktank/sharktank/models/grok/grok.py [hunk 2]
```diff
@@ -122,33 +123,32 @@ def prefill(
         self._assert_device(seq_block_ids)
         self._assert_device(*cache_state, dtype=self.activation_dtype)
         h = self.token_embedding(tokens)
-        h *= 78.38367176906169
+        h *= math.sqrt(h.shape[-1])
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/grok/grok.py:126]: This requires `import math`

Comment (dan-garvey) [reply sharktank/sharktank/models/grok/grok.py:126]: better as torch? I don't think math traces well

Comment (IanNod) [reply sharktank/sharktank/models/grok/grok.py:126]: Do we not have prefill here testing export? If not that needs to be added

File: sharktank/sharktank/models/grok/grok.py [hunk 3]
```diff
@@ -200,34 +200,34 @@ def decode(
...
+                h,
+                start_positions=start_positions,
+                embedding=self.attention_embedding,
+                embedding_batch_mask=embedding_batch_mask,
+                attention_mask=attention_mask,
+                cache_state=cache_state,
+                seq_block_ids=seq_block_ids,
+                xk_temp=xk_temp,
+                xv_temp=xv_temp,
+                softcap=30.0,
```
Comment (IanNod) [inline sharktank/sharktank/models/grok/grok.py:223]: would be good to have comments that identifying where these hard coded numbers come from

Comment (KyleHerndon) [reply sharktank/sharktank/models/grok/grok.py:223]: Would you want like a link to the llama.cpp or grok model github line? We're literally just copying from the publicly available implementation of grok. There's some logic relating to quantization behind it, but we didn't derive it ourselves so I don't know that we could give a "real" explanation.

File: sharktank/sharktank/models/grok/grok.py (hunk 1 above)
Comment (rsuderman) [reply sharktank/sharktank/models/grok/grok.py]: Fixed.

File: sharktank/sharktank/models/grok/grok.py (hunk 2 above)
Comment (rsuderman) [reply sharktank/sharktank/models/grok/grok.py]: We should try to make faked versions of each model that is locally exportable. @KyleHerndon can you look into making faked `theta` parameters for the model? It can just be a single layer with smallers tensors.

File: sharktank/sharktank/export_layer/export_moe.py [hunk 4]
```diff
@@ -46,12 +49,12 @@ def main():
...
     bs = args.batch_size
 
-    model = PreGatherMoeBlock(
+    model = MoeBlock(
         theta=make_moe_block_theta()("blk.0"),
         expert_count=8,
         expert_used_count=2,
         rms_epsilon=1e-5,
-        use_grok=args.use_grok,
+        moe_activation=F.gelu if args.use_grok else F.silu,
```
[the remark this answers is not in the document]
Comment (rsuderman) [reply sharktank/sharktank/export_layer/export_moe.py]: Done.

File: sharktank/sharktank/models/grok/grok.py (hunk 3 above)
Comment (rsuderman) [reply sharktank/sharktank/models/grok/grok.py]: Added a link to where we pull the hardcoded from.

Comment (IanNod) [reply sharktank/sharktank/models/grok/grok.py]: Yeah, link to the public implementations would be good. Just so someone not familiar with our codebase has some clue as to where these are coming from

File: sharktank/sharktank/models/grok/grok.py [hunk 5]
```diff
@@ -122,33 +123,32 @@ def prefill(
         self._assert_device(seq_block_ids)
         self._assert_device(*cache_state, dtype=self.activation_dtype)
         h = self.token_embedding(tokens)
-        h *= 78.38367176906169
+        h *= tanh.sqrt(h.shape[-1])
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/grok/grok.py:126]: You mean `torch.tanh`?

File: sharktank/sharktank/models/grok/grok.py [hunk 6]
```diff
@@ -123,7 +123,7 @@ def prefill(
         self._assert_device(seq_block_ids)
         self._assert_device(*cache_state, dtype=self.activation_dtype)
         h = self.token_embedding(tokens)
-        h *= math.sqrt(h.shape[-1])
+        h *= tanh.sqrt(h.shape[-1])
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/grok/grok.py:126]: You mean `torch.sqrt()`

Comment (archana-ramalingam) [reply sharktank/sharktank/models/grok/grok.py:126]: Also, everywhere we use `torch.sqrt()`, we need to make sure the input is in tensor form and not an `int`.

File: sharktank/sharktank/layers/paged_llama_attention_block.py [hunk 7]
```diff
@@ -156,7 +156,7 @@ def repeat_kv(x: torch.Tensor) -> torch.Tensor:
 
         attn_weights = ops.matmul(xq, keys.transpose(2, 3))
         if self.attention_scale is None:
-            attn_weights = attn_weights / math.sqrt(self.head_dim)
+            attn_weights = attn_weights / torch.sqrt(self.head_dim)
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/paged_llama_attention_block.py:159]: Remove `import math` as it's unused now.

File: sharktank/sharktank/models/grok/grok.py (hunk 2 above)
Comment (archana-ramalingam) [reply sharktank/sharktank/models/grok/grok.py]: [Here](https://github.com/nod-ai/SHARK-Platform/blob/main/sharktank/sharktank/models/llama/testing.py) are the theta generators for attention/FFN/MOE blocks. We already have attention and prefill tests, might need to add export component in there.

File: sharktank/sharktank/models/grok/grok.py [hunk 8]
```diff
@@ -200,34 +200,34 @@ def decode(
...
 
         h = self.token_embedding(tokens)
-        h *= 78.38367176906169
+        h *= torch.sqrt(h.shape[-1])
         self.trace_tensor("grok.token_embedding", h)
 
         # Iterate over attention blocks.
-        for block_idx, block in enumerate(self.attn_blocks):
+        for block_idx, (attn_block, moe_block) in enumerate(
+            self.attn_blocks, self.moe_blocks
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/grok/grok.py:208]: ` zip(self.attn_blocks, self.moe_blocks)`

File: sharktank/sharktank/layers/paged_llama_attention_block.py [hunk 9]
```diff
@@ -147,16 +154,16 @@ def repeat_kv(x: torch.Tensor) -> torch.Tensor:
...
+
         # Flash attention.
-        if not self.use_grok:
-            attn_weights = ops.matmul(xq, keys.transpose(2, 3)) / math.sqrt(
-                self.head_dim
-            )
-        elif self.use_grok:
-            attn_weights = ops.matmul(xq, keys.transpose(2, 3))
-            attn_weights = 30.0 * torch.tanh(
-                attn_weights * (0.08838834764831845 / 30.0)
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/paged_llama_attention_block.py:158]: Should we also hardcode the `attention_scale = 0.08838834764831845` in `grok.py`?

File: sharktank/sharktank/models/grok/grok.py [hunk 10]
```diff
@@ -200,34 +200,34 @@ def decode(
...
+                h,
+                start_positions=start_positions,
+                embedding=self.attention_embedding,
+                embedding_batch_mask=embedding_batch_mask,
+                attention_mask=attention_mask,
+                cache_state=cache_state,
+                seq_block_ids=seq_block_ids,
+                xk_temp=xk_temp,
+                xv_temp=xv_temp,
+                softcap=30.0,  # https://github.com/xai-org/grok-1/blob/7050ed204b8206bb8645c7b7bbef7252f79561b0/model.py#L864
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/grok/grok.py:223]: Does `softcap` belong to `PagedLlamaAttentionBlock()` and not `decode()`.

Comment (rsuderman) [reply sharktank/sharktank/models/grok/grok.py]: Whoops, should have put it in the initializer. Fixed.

File: sharktank/sharktank/layers/paged_llama_attention_block.py (hunk 9 above)
Comment (rsuderman) [reply sharktank/sharktank/layers/paged_llama_attention_block.py:158]: So the secret is `0.08838834764831845` is not a random hardcoded value, its actually `1.0 / sqrt(128)` and for `grok` `head_dim == 128`. So that normalization is already handled by the default attention scaling above.

File: sharktank/sharktank/models/grok/grok.py (hunk 6 above)
Comment (rsuderman) [reply sharktank/sharktank/models/grok/grok.py]: changed back to `math.sqrt`. For any of the sqrt of dims we need to use the math variant. All of these are constants though so this should be fine.

File: sharktank/sharktank/models/grok/grok.py (hunk 8 above)
Comment (rsuderman) [reply sharktank/sharktank/models/grok/grok.py]: Fixed.

**Review by IanNod — APPROVED** [review]
Looks good to me

**Review by archana-ramalingam — APPROVED** [review]
